api(download): Add saveAs helper

rwoll opened the pull request:
Closes #2523.

rwoll commented on test/download.spec.js:
```diff
@@ -57,6 +67,59 @@ describe('Download', function() {
+  it('should error when saving to non-existent user-specified path', async({persistentDirectory, browser, server}) => {
+    const page = await browser.newPage({ acceptDownloads: true });
+    await page.setContent(`<a href="${server.PREFIX}/download">download</a>`);
+    const [ download ] = await Promise.all([
+      page.waitForEvent('download'),
+      page.click('a')
+    ]);
+    const nonExistentUserPath = path.join(persistentDirectory, "does-not-exist","download.txt");
+    const { code, syscall, message } = await download.saveAs(nonExistentUserPath).catch(e => e);
+    expect(code).toBe('ENOENT');
+    expect(syscall).toBe('copyfile');
+    expect(message).toContain('no such file or directory');
+    expect(message).toContain('does-not-exist');
+    await page.close();
+  });
+  it('should error when saving with downloads disabled', async({persistentDirectory, browser, server}) => {
+    const page = await browser.newPage({ acceptDownloads: false });
+    await page.setContent(`<a href="${server.PREFIX}/download">download</a>`);
+    const [ download ] = await Promise.all([
+      page.waitForEvent('download'),
+      page.click('a')
+    ]);
+    const userPath = path.join(persistentDirectory, "download.txt");
+    const { message } = await download.saveAs(userPath).catch(e => e);
+    expect(message).toContain('Pass { acceptDownloads: true } when you are creating your browser context');
+    await page.close();
+  });
+  it('should error when saving after deletion', async({persistentDirectory, browser, server}) => {
+    const page = await browser.newPage({ acceptDownloads: true });
+    await page.setContent(`<a href="${server.PREFIX}/download">download</a>`);
+    const [ download ] = await Promise.all([
+      page.waitForEvent('download'),
+      page.click('a')
+    ]);
+    const userPath = path.join(persistentDirectory, "download.txt");
+    await download.delete();
+    const { message } = await download.saveAs(userPath).catch(e => e);
+    expect(message).toContain('Download already deleted. Save before deleting.');
+    await page.close();
+  });
```
I'm tempted to move 

```
    const page = await browser.newPage({ acceptDownloads: true });
    await page.setContent(`<a href="${server.PREFIX}/download">download</a>`);
    const [ download ] = await Promise.all([
      page.waitForEvent('download'),
      page.click('a')
    ]);
```

into a `beforeEach`. Thoughts?

rwoll commented:
@pavelfeldman Can you take a look at this and see if it aligns with what you had in mind in v1.3? I'm using this API in a personal project now and it's more convenient passing the download to a helper function.

pavelfeldman commented on docs/api.md:
```diff
@@ -3197,6 +3198,12 @@ Returns download error if any.
+- `path` <[string]> Path where the download should be saved. The directory structure MUST exist as `saveAs` will not create it.
+- returns: <[Promise]<void>>
+
+Returns path to the downloaded file in case of successful download.
```
Does not return anything...

rwoll replied:
🤦 Fixed! Thanks.

pavelfeldman commented on src/download.ts:
```diff
@@ -72,6 +72,15 @@ export class Download {
   }
 
+  async saveAs(path: string) {
+    const internalPath = await this.path();
```
You are racing the microtasks that are chained to `this.path()` in `saveAs` and `delete`. As a result, copy might happen after concurrently / after the unlink. I would rather store the request and await it in `_reportFinished` prior to calling `this._finishedCallback`.

rwoll replied:
I see the race and agree it needs to change, but I'm not completely following the requested change.

What happens if you call `saveAs` multiple times?

```
await Promise.all([dl.saveAs('foo'), dl.saveAs('bar')]) 
```

What happens if you call `saveAs` after `_reportFinished` is already run?

rwoll replied:
@pavelfeldman I made an attempt at it here: https://github.com/microsoft/playwright/pull/2872/commits/afe361ccef754ade7e19b0c50314310a2b04686e. I feel like I have one extra level of wrapping than needed. Also, is it ok that `_reportFinished` is now marked as `async`? The caller does not wait for it to resolve, but I think that's fine.

pavelfeldman reviewed (commented):
Sorry for the delay, this looks perfect, one review comment and it is fine to land. Great test coverage!

pavelfeldman commented:
@rwoll ^^

rwoll committed:
fix(download): Fix race in saveAs

rwoll committed:
test: add more saveAs tests

rwoll commented on test/download.spec.js:
```diff
@@ -57,6 +67,110 @@ describe('Download', function() {
     expect(fs.readFileSync(path).toString()).toBe('Hello world');
     await page.close();
   });
+  fit('should save to user-specified path', async({persistentDirectory, browser, server}) => {
```
```suggestion
  it('should save to user-specified path', async({persistentDirectory, browser, server}) => {
```

rwoll commented on test/download.spec.js:
```diff
@@ -57,6 +67,110 @@ describe('Download', function() {
+    expect(fs.readFileSync(userPath).toString()).toBe('Hello world');
+    await page.close();
+  });
+  fit('should save to user-specified path without updating original path', async({persistentDirectory, browser, server}) => {
```
```suggestion
  it('should save to user-specified path without updating original path', async({persistentDirectory, browser, server}) => {
```

rwoll committed:
test: remove test focus

pavelfeldman commented on src/download.ts:
```diff
@@ -95,8 +112,10 @@ export class Download {
-  _reportFinished(error?: string) {
+  async _reportFinished(error?: string) {
     this._failure = error || null;
+    for (const f of this._saveAsCallbacks)
```
I don't see where this is assigned to.

rwoll replied:
Oops. It was supposed to be this:

```
diff --git a/src/download.ts b/src/download.ts
index 1b3b67f0..8e6bf740 100644
--- a/src/download.ts
+++ b/src/download.ts
@@ -74,18 +74,20 @@ export class Download {
   }
 
   async saveAs(path: string) {
-    return new Promise(async (res, rej) => {
-      try {
-        const internalPath = await this.path();
-        if (internalPath === null)
-          throw new Error('Download not found on disk. Check download.failure() for details.');
-        if (this._deleted)
-          throw new Error('Download already deleted. Save before deleting.');
-        await util.promisify(fs.copyFile)(internalPath, path);
-        res();
-      } catch (error) {
-        rej(error);
-      }
+    return new Promise((res, rej) => {
+      this._saveAsCallbacks.push(async () => {
+        try {
+          const internalPath = await this.path();
+          if (internalPath === null)
+            throw new Error('Download not found on disk. Check download.failure() for details.');
+          if (this._deleted)
+            throw new Error('Download already deleted. Save before deleting.');
+          await util.promisify(fs.copyFile)(internalPath, path);
+          res();
+        } catch (error) {
+          rej(error);
+        }
+      })
     });
   }
```

I just tried the above patch and think it's now deadlocking. What did you have in mind to prevent the race?

pavelfeldman commented:
@rwoll ^^

rwoll committed:
fix(download): Fix race per-feedback

rwoll committed:
api(download): Add saveAs helper

Closes #2523.

rwoll commented:
@pavelfeldman I took another pass at it trying to incorporate your outline. Feel free to push changes directly as needed as I'm not sure when I'm going to have time to cycle back. I'm happy to have contributed tests, docs, and a partial implementation. :)

rwoll committed:
feat(rpc): pass failing RPC download tests

rwoll committed:
fix pdl definition

rwoll commented:
As of a554ce5ab75f76827a03f536c055215ec4a9a347, the saveAs tests hang for the RPC set of tests.

rwoll committed:
fix(download): add missing RPC await

rwoll committed:
test: relax expected error message struct

The strict assertions worked for the direct execution model, but with
the RPC layer, there is some error wrapping.

rwoll commented:
All checks are now passing (including the RPC tests)! 🟢 🚀 FWIW, I imagine the RPC versions of `download.saveAs` (and `download.path`) won't be used since it refers to server-side file locations. The RPC client consumer would probably want to do a client-side save as using stream and native file storage.)

pavelfeldman commented on src/download.ts:
```diff
@@ -72,6 +74,26 @@ export class Download {
+  }
+
+  async _saveAs(dlPath: string) {
+    if (!this._acceptDownloads)
```
nit: this can be in the `saveAs`

pavelfeldman commented on src/download.ts:
```diff
@@ -95,7 +117,26 @@ export class Download {
+  async _reportFinished(error?: string) {
+    if (error) {
+      for (const { reject } of this._saveAsRequests) {
+        if (!this._acceptDownloads)
```
Not needed here if you move the check into saveAs

pavelfeldman reviewed (approved):
Looks great! We'll follow to to support the rpc version!

pavelfeldman commented:
I'll merge this and follow up, those are minor comments. Great job and thank you!

rwoll commented:
Awesome! Thanks for your guidance and reviews throughout, @pavelfeldman!
